tools/topology/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+use std::collections::{BTreeMap, BTreeSet, VecDeque};
+
+use serde::Deserialize;
+
+pub use self::currency_definition::CurrencyDefinition;
+use self::{
+    inner_structure::{Channel, Currency, HostNetwork, IbcCurrency, NativeCurrency, Network},
+    symbol::Builder,
+};
+
+mod currency_definition;
+pub mod error;
+mod inner_structure;
+mod symbol;
+
+#[derive(Debug, Clone, Deserialize)]
+#[serde(from = "self::inner_structure::Raw")]
+pub struct Topology {
+    host_network: HostNetwork,
+    networks: BTreeMap<Box<str>, Network>,
+    channels: Box<[Channel]>,
+}
+
+impl Topology {
+    pub fn currency_definitions(
+        &self,
+        dex_network: &str,
+    ) -> Result<Box<[CurrencyDefinition]>, error::CurrencyDefinitions> {
+        let dex_currencies = &self
+            .networks
+            .get(dex_network)
+            .ok_or(error::CurrencyDefinitions::NonExistentDexNetwork)?
+            .currencies;
+
+        let channels = self.process_channels()?;
+
+        let host_to_dex_path =
+            Self::host_to_dex_path(&channels, &self.host_network.name, dex_network)?;
+
+        let mut currencies = vec![];
+
+        currencies.reserve_exact(dex_currencies.len());
+
+        dex_currencies
+            .iter()
+            .try_for_each(|(ticker, currency)| {
+                self.resolve_currency(dex_network, &channels, &host_to_dex_path, ticker, currency)
+                    .map(|currency| currencies.push(currency))
+            })
+            .map(|()| currencies.into_boxed_slice())
+            .map_err(Into::into)
+    }
+
+    fn host_to_dex_path<'r>(
```

Comment:
refuse to review functions longer than 20 loc, this one is more than 100!

Reply:
That's a fair criticism about the >100LoC, but I just want to be clear that not everything can be fit into <20LoC without passing 7+ parameters. 😉

Reply:
no comment